File: js/test/fixtures/acp-fake-server.mjs
#!/usr/bin/env node
// A minimal, scripted ACP (Agent Client Protocol) server for js/test/acp.test.ts.
// Built-ins only (node:readline over stdin/stdout) — no framework, no deps.
//
// Speaks JSON-RPC 2.0, one object per line, over its own stdin/stdout, the
// same shape a real ACP agent (devin acp, Gemini CLI, Zed's agents) uses.
// It is NOT a general ACP implementation — it is a deterministic script,
// selected with the ACP_SCENARIO env var, mirroring spikes/acp/fakeagent's
// scenario set by DESIGN (not by copying its Go code):
//
//   default    - trivial echo; used for the warm-session-reuse test. Replies
//                with "echo:<turnNumber>:<promptText>".
//   stale      - stateful session; answers based on the SUPERSEDES-ALL-PRIOR
//                marker when present, else a naive substring match against
//                remembered history (the failure the marker exists to avoid).
//   noisy      - emits agent_thought_chunk + tool-call narration interleaved
//                with a JSON payload split across two agent_message_chunks.
//   permission - sends session/request_permission (reject option first, then
//                an allow option) and waits for the reply before finishing
//                the turn, echoing back which optionId it received.
//   hang       - sends session/request_permission and then never answers the
//                session/prompt itself (models an unresponsive agent, so the
//                CLIENT's own safety-net timeout is what has to save it).
//   serialize  - tracks a `busy` flag; if a second session/prompt arrives
//                while busy, replies with an error instead of processing it
//                normally, so a client that fails to serialise turns is
//                caught red-handed.
//
// If ACP_RECORD_FILE is set, every session/new call appends its raw params
// as one JSON line to that file — used to assert the real client sends an
// absolute cwd + an mcpServers array (the -32602 trap real `devin acp`
// enforces).
import readline from "node:readline"
import fs from "node:fs"

const scenario = process.env.ACP_SCENARIO || "default"
const recordFile = process.env.ACP_RECORD_FILE

function send(msg) {
  process.stdout.write(JSON.stringify(msg) + "\n")
}

function sendNotification(method, params) {
  send({ jsonrpc: "2.0", method, params })
}

let nextServerReqId = 0
function sendRequest(method, params) {
  const id = `srv-${++nextServerReqId}`
  send({ jsonrpc: "2.0", id, method, params })
  return id
}

function reply(id, result) {
  send({ jsonrpc: "2.0", id, result })
}

function replyError(id, code, message) {
  send({ jsonrpc: "2.0", id, error: { code, message } })
}

const history = []
let promptN = 0
let busy = false
const permissionReplies = new Map() // id -> resolve()
const pendingPermissionResolvers = []

const rl = readline.createInterface({ input: process.stdin, terminal: false })

rl.on("line", (line) => {
  const trimmed = line.trim()
  if (!trimmed) return
  let msg
  try {
    msg = JSON.parse(trimmed)
  } catch {
    return
  }

  // A reply to a server-initiated request (session/request_permission)
  // arrives with an id we generated and no method.
  if (msg.method === undefined && msg.id !== undefined) {
    const resolver = permissionReplies.get(msg.id)
    if (resolver) {
      permissionReplies.delete(msg.id)
      resolver(msg)
    }
    return
  }

  switch (msg.method) {
    case "initialize":
      reply(msg.id, { protocolVersion: 1, agentCapabilities: { loadSession: false } })
      break

    case "session/new":
      if (recordFile) fs.appendFileSync(recordFile, JSON.stringify(msg.params ?? {}) + "\n")
      reply(msg.id, { sessionId: "sess-1" })
      break

    case "session/set_mode":
      reply(msg.id, {})
      break

    case "session/prompt": {
      const parts = (msg.params?.prompt ?? []).map((p) => p.text ?? "")
      const userText = parts.join("")
      promptN += 1
      history.push(userText)
      handlePrompt(msg.id, msg.params?.sessionId, userText)
      break
    }

    case "session/cancel":
      reply(msg.id, {})
      break

    default:
      if (msg.id !== undefined) reply(msg.id, {})
  }
})

function handlePrompt(id, sessionId, userText) {
  switch (scenario) {
    case "stale":
      handleStale(id, userText)
      return
    case "noisy":
      handleNoisy(id, sessionId, userText)
      return
    case "permission":
      handlePermission(id, sessionId, userText)
      return
    case "hang":
      // Ask for permission (the auto-answering client answers it
      // instantly) but never reply to session/prompt itself — only the
      // client's own safety-net timeout can end this turn.
      sendRequest("session/request_permission", {
        sessionId,
        options: [{ optionId: "allow-once", kind: "allow_once", name: "Allow" }],
      })
      return
    case "serialize":
      handleSerialize(id, sessionId, userText)
      return
    default:
      sendNotification("session/update", {
        sessionId,
        update: { sessionUpdate: "agent_message_chunk", content: { type: "text", text: `echo:${promptN}:${userText}` } },
      })
      reply(id, { stopReason: "end_turn" })
  }
}

const MARKER = "SUPERSEDES-ALL-PRIOR:"

function handleStale(id, userText) {
  let answer
  const idx = userText.indexOf(MARKER)
  if (idx >= 0) {
    const fresh = userText.slice(idx + MARKER.length).trim()
    answer = `FRESH-ANSWER-TO:${fresh}`
  } else {
    let matched = history[0]
    for (const h of history) {
      if (userText.includes(h)) {
        matched = h
        break
      }
    }
    answer = `STALE-ANSWER-TO:${matched}`
  }
  sendNotification("session/update", {
    sessionId: "sess-1",
    update: { sessionUpdate: "agent_message_chunk", content: { type: "text", text: answer } },
  })
  reply(id, { stopReason: "end_turn" })
}

function handleNoisy(id, sessionId, userText) {
  const chunks = [
    { sessionUpdate: "agent_thought_chunk", content: { type: "text", text: "Let me think about this... " } },
    { sessionUpdate: "tool_call", toolCallId: "t1", title: "reading files", status: "in_progress" },
    { sessionUpdate: "agent_message_chunk", content: { type: "text", text: '{"answer":' } },
    { sessionUpdate: "tool_call_update", toolCallId: "t1", status: "completed" },
    { sessionUpdate: "agent_thought_chunk", content: { type: "text", text: "now double-checking... " } },
    { sessionUpdate: "agent_message_chunk", content: { type: "text", text: `${JSON.stringify(userText)}}` } },
  ]
  for (const c of chunks) sendNotification("session/update", { sessionId, update: c })
  reply(id, { stopReason: "end_turn" })
}

function handlePermission(id, sessionId, userText) {
  const reqId = sendRequest("session/request_permission", {
    sessionId,
    options: [
      { optionId: "reject", kind: "reject_once", name: "Reject" },
      { optionId: "allow-once", kind: "allow_once", name: "Allow" },
    ],
  })
  permissionReplies.set(reqId, (m) => {
    const optionId = m.result?.outcome?.optionId ?? "none"
    sendNotification("session/update", {
      sessionId,
      update: { sessionUpdate: "agent_message_chunk", content: { type: "text", text: `PERMITTED:${optionId}` } },
    })
    reply(id, { stopReason: "end_turn" })
  })
}

function handleSerialize(id, sessionId, userText) {
  if (busy) {
    replyError(id, -32001, "reentrant session/prompt received while a prior turn was still in flight")
    return
  }
  busy = true
  setTimeout(() => {
    sendNotification("session/update", {
      sessionId,
      update: { sessionUpdate: "agent_message_chunk", content: { type: "text", text: `done:${userText}` } },
    })
    reply(id, { stopReason: "end_turn" })
    busy = false
  }, 50)
}

File: python/tests/fixtures/acp_fake_server.py
#!/usr/bin/env python3
"""A scripted fake ACP (Agent Client Protocol) server for hermetic Python tests of
``toolnexus.acp``.

Mirrors ``spikes/acp/fakeagent/main.go``'s scenario set and protocol-handling
*design* (JSON-RPC 2.0, one object per line, over its own stdin/stdout, with
server-initiated requests such as ``session/request_permission`` interleaved on the
same stream) — reimplemented here in Python, stdlib only, for this port's own test
suite. Not a general ACP implementation; a deterministic script selected with
``--scenario``.

Scenarios:
  default    - trivial instant echo: one ``agent_message_chunk`` per turn.
  stale      - stateful per-session history; without the client's
               ``SUPERSEDES-ALL-PRIOR`` marker it answers whichever earlier prompt
               is a substring of the current one (stale); with the marker present
               it answers the fresh text named by the marker.
  hang       - never replies to ``session/prompt`` and never asks permission — used
               to prove the client's own safety-net timeout is what ends the wait.
  permission - sends ``session/request_permission`` mid-turn and blocks until the
               client answers, then echoes the chosen ``optionId`` back in the
               final ``agent_message_chunk``.
  noisy      - interleaves ``agent_thought_chunk`` and ``tool_call``/
               ``tool_call_update`` notifications around the real
               ``agent_message_chunk`` content, split across >= 2 notifications.
  serialize  - records reentrancy (a "busy" flag) if a second ``session/prompt``
               arrives while one is already being handled on this session — proves
               the client serialises turns.

``diagnostics/get`` is a JSON-RPC method that is NOT part of real ACP: it returns
this process's internal counters as the reply ``result``, so a test can assert
things like "session/new happened exactly once" or "no reentrant session/prompt was
observed" without depending on timing.
"""
from __future__ import annotations

import json
import sys
import threading
import time

_STDOUT_LOCK = threading.Lock()


def _send(obj: dict) -> None:
    line = json.dumps(obj)
    with _STDOUT_LOCK:
        sys.stdout.write(line + "\n")
        sys.stdout.flush()


def main() -> None:
    args = sys.argv[1:]
    scenario = "default"
    if "--scenario" in args:
        scenario = args[args.index("--scenario") + 1]

    state_lock = threading.Lock()
    state = {
        "session_new_count": 0,
        "prompt_count": 0,
        "busy": False,
        "reentrant_detected": False,
        "history": [],
        "last_option_id": None,
        "received_cwd": None,
        "received_mcp_servers": None,
    }
    next_srv_id = [0]
    perm_replies: dict[str, dict] = {}
    perm_cv = threading.Condition()

    def send_request(method: str, params: dict) -> str:
        next_srv_id[0] += 1
        rid = f"srv-{next_srv_id[0]}"
        _send({"jsonrpc": "2.0", "id": rid, "method": method, "params": params})
        return rid

    def send_notification(method: str, params: dict) -> None:
        _send({"jsonrpc": "2.0", "method": method, "params": params})

    def reply(msg_id, result) -> None:
        _send({"jsonrpc": "2.0", "id": msg_id, "result": result})

    def wait_for_perm_reply(rid: str) -> dict:
        with perm_cv:
            while rid not in perm_replies:
                perm_cv.wait()
            return perm_replies.pop(rid)

    def send_message_chunk(session_id: str, text: str) -> None:
        send_notification(
            "session/update",
            {
                "sessionId": session_id,
                "update": {"sessionUpdate": "agent_message_chunk", "content": {"type": "text", "text": text}},
            },
        )

    def handle_prompt(msg_id, session_id: str, user_text: str) -> None:
        with state_lock:
            state["prompt_count"] += 1
            state["history"].append(user_text)
            history_snapshot = list(state["history"])

        if scenario == "hang":
            # Never reply, never ask permission. The client's own safety-net
            # timeout is the only thing that ever ends its wait.
            return

        if scenario == "permission":
            rid = send_request(
                "session/request_permission",
                {
                    "sessionId": session_id,
                    "options": [
                        {"optionId": "reject", "kind": "reject_once", "name": "Reject"},
                        {"optionId": "allow-once", "kind": "allow_once", "name": "Allow"},
                    ],
                },
            )
            answer = wait_for_perm_reply(rid)
            outcome = (answer.get("result") or {}).get("outcome") or {}
            option_id = outcome.get("optionId")
            with state_lock:
                state["last_option_id"] = option_id
            send_message_chunk(session_id, f"PERMITTED:{option_id}")
            reply(msg_id, {"stopReason": "end_turn"})
            return

        if scenario == "noisy":
            send_notification(
                "session/update",
                {
                    "sessionId": session_id,
                    "update": {"sessionUpdate": "agent_thought_chunk", "content": {"type": "text", "text": "Let me think about this... "}},
                },
            )
            send_notification(
                "session/update",
                {"sessionId": session_id, "update": {"sessionUpdate": "tool_call", "toolCallId": "t1", "title": "reading files", "status": "in_progress"}},
            )
            send_message_chunk(session_id, '{"answer":')
            send_notification(
                "session/update",
                {"sessionId": session_id, "update": {"sessionUpdate": "tool_call_update", "toolCallId": "t1", "status": "completed"}},
            )
            send_notification(
                "session/update",
                {
                    "sessionId": session_id,
                    "update": {"sessionUpdate": "agent_thought_chunk", "content": {"type": "text", "text": "now double-checking the number... "}},
                },
            )
            send_message_chunk(session_id, f'"{user_text}"}}')
            reply(msg_id, {"stopReason": "end_turn"})
            return

        if scenario == "stale":
            marker = "SUPERSEDES-ALL-PRIOR:"
            idx = user_text.find(marker)
            if idx >= 0:
                fresh = user_text[idx + len(marker):].strip()
                answer = f"FRESH-ANSWER-TO:{fresh}"
            else:
                matched = history_snapshot[0] if history_snapshot else ""
                for h in history_snapshot:
                    if h and h in user_text:
                        matched = h
                        break
                answer = f"STALE-ANSWER-TO:{matched}"
            send_message_chunk(session_id, answer)
            reply(msg_id, {"stopReason": "end_turn"})
            return

        if scenario == "serialize":
            with state_lock:
                if state["busy"]:
                    state["reentrant_detected"] = True
                state["busy"] = True
            time.sleep(0.15)
            send_message_chunk(session_id, f"echo:{user_text}")
            with state_lock:
                state["busy"] = False
            reply(msg_id, {"stopReason": "end_turn"})
            return

        # default
        send_message_chunk(session_id, f"echo:{user_text}")
        reply(msg_id, {"stopReason": "end_turn"})

    for raw_line in sys.stdin:
        line = raw_line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            continue

        method = msg.get("method")
        msg_id = msg.get("id")

        if not method and msg_id is not None:
            # A reply to one of THIS process's server-initiated requests
            # (session/request_permission).
            with perm_cv:
                perm_replies[msg_id] = msg
                perm_cv.notify_all()
            continue

        if method == "initialize":
            reply(msg_id, {"protocolVersion": 1, "agentCapabilities": {"loadSession": False}})
        elif method == "session/new":
            params = msg.get("params") or {}
            with state_lock:
                state["session_new_count"] += 1
                count = state["session_new_count"]
                state["received_cwd"] = params.get("cwd")
                state["received_mcp_servers"] = params.get("mcpServers")
            reply(
                msg_id,
                {
                    "sessionId": "sess-1",
                    "sessionNewCount": count,
                    "receivedCwd": params.get("cwd"),
                    "receivedMcpServers": params.get("mcpServers"),
                },
            )
        elif method == "session/set_mode":
            reply(msg_id, {})
        elif method == "session/prompt":
            params = msg.get("params") or {}
            session_id = params.get("sessionId")
            text = "".join(p.get("text", "") for p in (params.get("prompt") or []))
            threading.Thread(target=handle_prompt, args=(msg_id, session_id, text), daemon=True).start()
        elif method == "session/cancel":
            reply(msg_id, {})
        elif method == "diagnostics/get":
            with state_lock:
                reply(msg_id, dict(state))
        else:
            if msg_id is not None:
                reply(msg_id, {})


if __name__ == "__main__":
    main()
